python/lib/client/dmod/client/request_clients.py

```diff
@@ -0,0 +1,206 @@
+from abc import ABC, abstractmethod
+from dmod.communication import InternalServiceClient, MaasRequestClient, ManagementAction
+from dmod.communication.client import R
+from dmod.communication.dataset_management_message import DatasetManagementMessage, DatasetManagementResponse, \
+    MaaSDatasetManagementMessage, MaaSDatasetManagementResponse
+from dmod.core.meta_data import DataCategory
+from pathlib import Path
+from typing import Generic, List, Optional, TypeVar, Type
+
+#import logging
+#logger = logging.getLogger("gui_log")
+
+
+DATA_RESPONSE = TypeVar("DATA_RESPONSE", bound=DatasetManagementResponse)
+
+
+class DatasetClient(Generic[DATA_RESPONSE], ABC):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.last_response = None
+
+    @abstractmethod
+    def _parse_list_of_dataset_names_from_response(self, response: DATA_RESPONSE) -> List[str]:
+        pass
+
+    @abstractmethod
+    async def create_dataset(self, name: str, category: DataCategory) -> bool:
+        pass
+
+    @abstractmethod
+    async def list_datasets(self, category: Optional[DataCategory] = None) -> List[str]:
+        pass
+
+    @abstractmethod
+    async def upload_to_dataset(self, dataset_name: str, paths: List[Path]) -> bool:
+        pass
+
+
+class DatasetInternalClient(DatasetClient[DatasetManagementResponse],
+                            InternalServiceClient[DatasetManagementMessage, DatasetManagementResponse]):
+
+    @classmethod
+    def get_response_subtype(cls) -> Type[R]:
+        return DatasetManagementResponse
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def _parse_list_of_dataset_names_from_response(self, response: DatasetManagementResponse) -> List[str]:
+        # TODO: *************************************
+        # TODO: how to parse the response for a list of dataset names?
+        pass
+
+    async def create_dataset(self, name: str, category: DataCategory) -> bool:
+        # TODO: (later) consider also adding param for data to be added
+        request = DatasetManagementMessage(action=ManagementAction.CREATE, dataset_name=name, category=category)
+        self.last_response = await self.async_make_request(request)
+        return self.last_response is not None and self.last_response.success
+
+    async def list_datasets(self, category: Optional[DataCategory] = None) -> List[str]:
+        action = ManagementAction.LIST_ALL if category is None else ManagementAction.SEARCH
+        request = DatasetManagementMessage(action=action, category=category)
+        self.last_response = await self.async_make_request(request)
+        return self._parse_list_of_dataset_names_from_response(self.last_response)
```

Comment:
so wouldn't this technically return `None` since the function is "unimplemented" with `pass`?

Reply:
You are correct, and it would have probably been a better practice to have that function return an empty list.  In the future, I will try to remember this and do something along those lines instead.

This one I am going to leave as-is, in the interest of reducing branch merging complexity.  An implementation will come along in another PR branch before too much longer.